File: baekjoon/python/9663.py
import sys
input=sys.stdin.readline

n=int(input())
arr=[0]*n # arr[i]=x: i번째(i행) 퀸은 x번째 열에 위치한다.
ans=0

# 퀸은 각 행 또는 열, 대각선에 하나만 위치해야 한다.
def cond(k):
    for i in range(k):
        if arr[i]==arr[k] or abs(arr[i]-arr[k])==abs(i-k):
            return False
        
    return True

def dfs(k):
    if(n==k):
        global ans
        ans+=1
        return
    
    for i in range(n):
        arr[k]=i
        if cond(k):
            dfs(k+1)
    
dfs(0)
print(ans)

File: baekjoon/cpp/solutions/9663.cpp
#include <iostream>

using namespace std;

int n;
int col[16]; // 퀸은 각 행 또는 열에 하나씩 있어야 하며 대각선으로 만나도록 배치하면 안 된다.
int ans = 0;

bool cond(int k) {
	for (int i = 0;i < k;i++) {
		if (col[i] == col[k] || abs(col[k] - col[i]) == k - i) {
			return false;
		}
	}
	return true;
}

void bt(int k) {
	if (n == k) {
		ans++;
		return;
	}

	for (int i = 0;i < n;i++) {
		col[k] = i;
		if (cond(k))
			bt(k + 1);
	}
}

int main() {
	cin >> n;

	bt(0);
	cout << ans;
}